packages/plugin-ext/src/plugin/command-registry.ts

```diff
@@ -123,3 +130,57 @@ export class CommandRegistryImpl implements CommandRegistryExt {
         this.argumentProcessors.push(processor);
     }
 }
+
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+// copied and modified from https://github.com/microsoft/vscode/blob/1.37.1/src/vs/workbench/api/common/extHostCommands.ts#L217-L259
+export class CommandsConverter {
+
+    private readonly safeCommandId: string;
+    private readonly commands: CommandRegistryImpl;
+    private readonly commandsMap = new Map<number, theia.Command>();
+    private handle = 0;
+    private isCommandRegistered: boolean;
+
+    constructor(commands: CommandRegistryImpl) {
+        this.safeCommandId = `theia_safe_cmd_${Date.now().toString()}`;
+        this.commands = commands;
+        this.isCommandRegistered = false;
+    }
+
+    /**
+     * Convert to a command that can be safely passed over JSON-RPC.
+     */
+    toSafeCommand(command: theia.Command, disposables: DisposableCollection): theia.Command {
+        if (!this.isCommandRegistered) {
+            this.commands.registerCommand({ id: this.safeCommandId }, this.executeSafeCommand, this);
+            this.isCommandRegistered = true;
+        }
+
+        const result: theia.Command = {};
+        Object.assign(result, command);
+
+        if (command.command && command.arguments && command.arguments.length > 0) {
```

Comment:
If I understand this correctly, this will make it impossible to invoke commands registered in one extension host from a different extension host (because the arguments are not passed through the front end). While this is a non-issue in Theia proper, it is a breaking change for Che plugins. Maybe we should run this change by the chedev mailing list first.

Reply:
It won't affect commands without arguments. This logic just re-register commands that have arguments. I don't see any reasons why a command won't be invoked.

Reply:
Ok, I see that "executeSafeCommand" delegates to the command registry for execution. So plugins have the option to have "unsafe" arguments (and execute locally) or safe arguments that will go through the "executeCommand" API call back to the front end.

Reply:
Actually commands will be executed with real arguments that was saved in the plugin side:
https://github.com/theia-ide/theia/blob/191bb6d02a841fe1f448809bfa1c64f032d00edb/packages/plugin-ext/src/plugin/command-registry.ts#L179
if the command was registered by the plugin it will be executed locally:
https://github.com/theia-ide/theia/blob/191bb6d02a841fe1f448809bfa1c64f032d00edb/packages/plugin-ext/src/plugin/command-registry.ts#L98-L99
Safe command is needed when the tree-view passes items to the front-end to prevent `JSON circular structure`error: https://github.com/theia-ide/theia/blob/191bb6d02a841fe1f448809bfa1c64f032d00edb/packages/plugin-ext/src/plugin/tree/tree-views.ts#L263